### PR TITLE
Scrapper, Sofia Nurtdinova - 22FPL2

### PR DESCRIPTION
**Pull request by sofianurtdinova**

### CODE REVIEW

**Comment by demid5111** [comment]
nothing to review

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -127,11 +233,23 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
         Returns:
             str: Url from HTML
         """
+        url = ''
+        links = article_bs.find_all(class_='afisha-page')
+        for link in links:
+            url = link.find('a').get('href')
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:242]: fixed

File: lab_5_scrapper/scrapper_config.json [hunk 2]
```diff
@@ -1,9 +1,18 @@
...
-    "timeout": 0,
-    "should_verify_certificate": true,
+    "seed_urls": [
+        "https://ti71.ru/news/afisha/?rubric=%D0%92%D1%8B%D1%81%D1%82%D0%B0%D0%B2%D0%BA%D0%B8",
+        "https://ti71.ru/news/afisha/?rubric=%D0%94%D0%B5%D1%82%D1%8F%D0%BC",
+        "https://ti71.ru/news/afisha/?rubric=%D0%A4%D0%B5%D1%81%D1%82%D0%B8%D0%B2%D0%B0%D0%BB%D0%B8",
+        "https://ti71.ru/news/afisha/?rubric=%D0%9A%D0%B8%D0%BD%D0%BE",
+        "https://ti71.ru/news/afisha/?rubric=%D0%9B%D0%B5%D0%BA%D1%86%D0%B8%D0%B8",
+        "https://ti71.ru/news/afisha/?rubric=%D0%AD%D0%BA%D1%81%D0%BA%D1%83%D1%80%D1%81%D0%B8%D0%B8"],
+    "headers" : {"Cookie" : "stg_returning_visitor=Mon%2C%2008%20Apr%202024%2011:26:31%20GMT; stg_last_interaction=Tue%2C%2016%20Apr%202024%2017:27:50%20GMT",
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper_config.json]: Thank you, I suppose, fixed

File: lab_5_scrapper/scrapper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (demid5111) [reply lab_5_scrapper/scrapper.py:242]: where?

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -176,6 +308,12 @@ def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> No
         Args:
             article_soup (bs4.BeautifulSoup): BeautifulSoup instance
         """
+        title = article_soup.find('h1', class_='news-detail__title')
+        self.article.title = title.text
+        date = article_soup.find('p', class_='news-detail__date date')
+        self.article.date = self.unify_date_format(date.text)
+        topic = article_soup.find('p', class_='news-detail__rubric tag tag--large')
+        self.article.topics = topic.text
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:316]: should not be topics a list?

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -187,6 +325,25 @@ def unify_date_format(self, date_str: str) -> datetime.datetime:
...
+            "августа": "Aug",
+            "сентября": "Sep",
+            "октября": "Oct",
+            "ноября": "Nov",
+            "декабря": "Dec"
+        }
+        for rus, eng in ruen_months.items():
+            if rus in date_str:
+                date_res = date_str.replace(rus, eng)
+        return datetime.datetime.strptime(date_res, '%d-%m-%Y %H:%M:%S')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:346]: 🎖️ very nice!

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -195,6 +352,11 @@ def parse(self) -> Union[Article, bool, list]:
         Returns:
             Union[Article, bool, list]: Article instance
         """
+        response = make_request(self.full_url, self.config)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:355]: you are so sure that request is always successful..

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -204,12 +366,26 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
 
 
 def main() -> None:
     """
     Entrypoint for scrapper module.
     """
+    conf = Config(CRAWLER_CONFIG_PATH)
+    crawler = Crawler(conf)
+    crawler.find_articles()
+    prepare_environment(ASSETS_PATH)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:381]: too late. start execution from preparing environment

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -204,12 +366,26 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+
+    for ind, url in enumerate(crawler.urls):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:383]: call ind as i, be more conventional

File: requirements.txt [hunk 8]
```diff
@@ -1 +1,3 @@
-
+beautifulsoup4 == 4.12.3
+lxml == 5.2.1
+requests==2.31.0
```
Comment (demid5111) [inline requirements.txt:3]: .

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -26,11 +89,38 @@ def _extract_config_content(self) -> ConfigDTO:
...
+                         encoding=conf["encoding"],
+                         timeout=conf["timeout"],
+                         should_verify_certificate=conf["should_verify_certificate"],
+                         headless_mode=conf["headless_mode"])
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        with open(self.path_to_config, "r", encoding="utf-8") as file:
```
[the remark this answers is not in the document]
Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: **reply to my comments - like "OK", "fixed", etc. it takes a lot of time understanding whether you tried to address or just ignored.**

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -18,6 +75,16 @@ def __init__(self, path_to_config: pathlib.Path) -> None:
...
+        self.path_to_config = path_to_config
+        self.conf_dto = self._extract_config_content()
+        self._validate_config_content()
+        self._seed_urls = self.conf_dto.seed_urls
+        self._num_articles = self.conf_dto.total_articles
+        self._headers = self.conf_dto.headers
+        self._encoding = self.conf_dto.encoding
+        self._timeout = self.conf_dto.timeout
+        self._should_verify_certificate = self.conf_dto.should_verify_certificate
+        self._headless_mode = self.conf_dto.headless_mode
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:87]: as far as i understand they give faster access with using direct attributes but I suppose hypothetically we could write in getters just get.conf_dto.(optional)... hope I got it right

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -26,19 +93,46 @@ def _extract_config_content(self) -> ConfigDTO:
...
+        with open(self.path_to_config, "r", encoding="utf-8") as file:
+            conf = json.load(file)
+        return ConfigDTO(**conf)
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
 
+        if not (isinstance(self.conf_dto.seed_urls, list) and all(
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:105]: fixed

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -26,19 +93,46 @@ def _extract_config_content(self) -> ConfigDTO:
...
+                re.match(r"(https)?://\w+\.", seed_url) for seed_url in self.conf_dto.seed_urls)):
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:106]: fixed

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:355]: hoped for the best... fixed

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:381]: fixed

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -127,11 +236,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
         Find articles.
         """
+        urls = []
+        while len(urls) < self.config.get_num_articles():
+            for seed_url in self.get_search_urls():
+                response = make_request(seed_url, self.config)
+                if response.status_code == 200:
+                    found = BeautifulSoup(response.text, 'lxml')
+                    urls.append(self._extract_url(found))
+        self.urls.append(urls)
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:256]: it planned to be "extend"...

File: lab_5_scrapper/scrapper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py]: I hope I fixed it now, but I also feel disorientated and like on the fence... please, let me know if I did it wrong again, I'll try to change sth

**Comment by sofianurtdinova** [comment]
> nice, especially dates management! continue towards 10 and make sure CI is green at least for 8

Thank you! I made it green for mark 8, will do my best to try to implement recursive crawler

File: requirements.txt [hunk 14]
```diff
@@ -1 +1,3 @@
-
+beautifulsoup4==4.12.3
+lxml==5.2.1
+requests==2.31.0
```
Comment (demid5111) [inline requirements.txt:3]: newline

File: lab_5_scrapper/scrapper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: it is right direction, but overcomplicated in logic. simplify

File: lab_5_scrapper/scrapper_config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (demid5111) [reply lab_5_scrapper/scrapper_config.json]: 🏅

**Review by demid5111 — CHANGES_REQUESTED** [review]
polish for 8. only minor comments are left

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -169,6 +290,7 @@ def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
             article_soup (bs4.BeautifulSoup): BeautifulSoup instance
         """
 
+
```
[the remark this answers is not in the document]
Comment (sofianurtdinova) [reply lab_5_scrapper/scrapper.py:306]: fixed

File: requirements.txt (hunk 14 above)
Comment (sofianurtdinova) [reply requirements.txt]: changed

**Pull request closed without merging**